README.md
# HTMLGraphics metric

Retrieves metric value from Grafana's [data interface](https://grafana.com/docs/grafana/latest/packages_api/data/paneldata/).

## Usage

The main difference between MetricValue and MetricData is that MetricValue is a single value while MetricData is an object

```ts
type MetricValue = unknown;

interface TimeData {
  first?: unknown;
  last?: unknown;
}

interface MetricData {
  calcs: { [key: string]: unknown };
  time: TimeData;
  hasData: boolean;
}
```

### getMetricValueFromExpression

Evaluates a metric expression

```ts
import { getMetricValueFromExpression } from "@gapit/grafana-metric";

// random-metric = 100

getMetricValueFromExpression("'random-metric' * 2"); // Returns 200
getMetricValueFromExpression("Math.sqrt('random-metric')"); // Returns 10
```

### getShowcaseMetricValue

Generate a random number

```ts
import { getShowcaseMetricValue } from "@gapit/grafana-metric";

getShowcaseMetricValue(); // Returns a value between 0 and 1000 with 2 decimals
getShowcaseMetricValue({ range: { min: x, max: y }, decimals: z }); // Return a random value between x and y with z decimals.
```

### getMetricValueFromName

Gets a metric value by name/alias

```ts
import { getMetricValueFromName } from "@gapit/grafana-metric";

// metric-name = 100

getMetricValueFromName("metric-name"); // Returns 100
getMetricValueFromName("non-existing-metric"); // Returns null
getMetricValueFromName("non-existing-metric", { noDataValue: "No data" }); // Returns "No data"
```

### getMetricValue

Function provides calculations for grafana queries.

```ts
import { getMetricValue } from "@gapit/grafana-metric";

// metric-name = 100

// From name
getMetricValue("metric-name"); // Returns 100

// No data
getMetricValue("non-existing-metric"); // Returns null
getMetricValue("non-existing-metric", { noDataValue: "No data" }); // Returns "No data"

// Evaluation string
getMetricValue("'metric-name' * 2"); // Returns 200
getMetricValue("Math.sqrt('metric-name')"); // Returns 10

// Showcase
getMetricValue("metric-name", { showcase: true }); // Returns a random value between 0 and 1000.
getMetricValue("metric-name", { showcase: true, range: { min: 0, max: 10 } }); // Returns random value between 1-10.
getMetricValue("metric-name", {
  showcase: true,
  range: { min: 0, max: 10 },
  decimals: 4,
}); // Returns random value between 1-10 with 4 decimals.
```

### getMetricDataFromExpression

Evaluates a metric expression

### getShowcaseMetricData

Generates a random metric data object

### getMetricDataFromName

Gets metric data by name/alias

### getMetricData

Gets metric data


CHANGELOG.md
# Changelog for grafana-metric

## v1.2.0 (2023/04/27)

### Features / enhancements

- Add method to filter by series name, field name and multiple labels [#35](https://github.com/gapitio/grafana-metric/pull/35)

## v1.1.1 (2022/01/07)

### Features / enhancements

- Check if field contains displayName if name fails [#29](https://github.com/gapitio/grafana-metric/pull/29)

## v1.1.0 (2021/08/03)

### Features / enhancements

- Use values if calcs is missing [#28](https://github.com/gapitio/grafana-metric/pull/28)

## v1.0.1 (2021/08/03)

### Bug fixes

- Fix wrong time returned [#27](https://github.com/gapitio/grafana-metric/pull/27)

## v1.0.0 (2021/03/05)

### Features / enhancements

- Add reducer ids [#13](https://github.com/gapitio/grafana-metric/pull/13)
- Add metricData which retrieves multiple calcs (last, min, max, first, ETC) and time [#18](https://github.com/gapitio/grafana-metric/pull/18)

## v0.2.0 (2021/02/24)

### Features / enhancements

- Add get metric from field names [#5](https://github.com/gapitio/grafana-metric/pull/5)
- Add get metric from label [#6](https://github.com/gapitio/grafana-metric/pull/6)
- Export unused functions (getFieldByName, getSeriesByName, getValueField) [#10](https://github.com/gapitio/grafana-metric/pull/10)

## v0.1.0 (2021/02/05)

### Features / enhancements

- Add getEvaluationString which evaluates a string with metric names
- Add better descriptions and examples

## v0.0.2 (2021/02/01)

### Bug fixes

- Fix missing noDataValue parameter

## v0.0.1 (2020/11/19)

### Features / enhancements

- Add getShowcaseMetricValue which returns a value between a set range of numbers with a set amount of decimals.
- Add getMetricValue which decides what function to run.

## v0.0.0 (2020/11/06)

### Features / enhancements

- Add getMetricValueByName which retrieves last metric value based off of the name provided.


src/utils/getMetricValue.ts
import { getEvaluatedString } from "./getEvaluatedString";
import { getMetricValueByName } from "./getMetricValueByName";
import { getShowcaseMetricValue } from "./getShowcaseMetricValue";

/**
 * Function provides calculations for grafana queries
 *
 * @example
 *
 * // metric-name = 100
 *
 * // By name
 * getMetricValue("metric-name") // Returns 100
 *
 * // No data
 * getMetricValue("non-existing-metric") // Returns null
 * getMetricValue("non-existing-metric", false, [0, 10], 2, "No data") // Returns "No data"
 *
 * // Evaluation string
 * getMetricValue("'metric-name' * 2") // Returns 200
 * getMetricValue("Math.sqrt('metric-name')") // Returns 10
 *
 * // Showcase
 * getMetricValue("metric-name", true); // Returns a random value between 0 and 1000.
 * getMetricValue("metric-name", true, [1, 10]); // Returns random value between 1-10.
 * getMetricValue("metric-name", true, [1, 10], 4); // Returns random value between 1-10 with 4 decimals.
 *
 * @param metric - String for alias used in grafana query.
 * @param showcase - Decides if values are randomly generated.
 * @param range - Range of values to return (only with showcase)
 * @param decimals - Amount of decimals returned (only with showcase)
 * @param noDataValue - Return value when no data is found.
 */
export function getMetricValue(
  metric: string,
  showcase = false,
  range?: [number, number],
  decimals?: number,
  noDataValue: unknown = null
): unknown {
  if (showcase) {
    return getShowcaseMetricValue({ range, decimals });
  } else if (metric.includes('"') || metric.includes("'")) {
    return getEvaluatedString(metric);
  }

  return getMetricValueByName(metric, { noDataValue });
}
